near-sdk/src/store/vec/mod.rs

```diff
@@ -0,0 +1,453 @@
+                            // Value was modified, write the updated value to storage
+                            env::storage_write(&key, &Self::serialize_element(modified));
+                        }
+                        None => {
+                            // Element was removed, clear the storage for the value
+                            env::storage_remove(&key);
+                        }
+                    }
+
+                    // Update state of flushed state as cached, to avoid duplicate writes/removes
+                    // while also keeping the cached values in memory.
+                    v.replace_state(EntryState::Cached);
+                }
+            }
+        }
+    }
+
+    /// Sets a value at a given index to the value provided. This does not shift values after the
+    /// index to the right.
+    pub fn set(&mut self, index: u32, value: T) {
+        if index >= self.len() {
+            env::panic(ERR_INDEX_OUT_OF_BOUNDS);
+        }
+
+        let entry = self.cache.get_mut(index);
+        match entry.get_mut() {
+            Some(entry) => *entry.value_mut() = Some(value),
+            None => {
+                let _ = entry.set(CacheEntry::new_modified(Some(value)));
+            }
+        }
+    }
+
+    fn serialize_element(element: &T) -> Vec<u8> {
+        element.try_to_vec().unwrap_or_else(|_| env::panic(ERR_ELEMENT_SERIALIZATION))
+    }
+
+    /// Appends an element to the back of the collection.
+    pub fn push(&mut self, element: T) {
+        if self.len() >= u32::MAX {
+            env::panic(ERR_INDEX_OUT_OF_BOUNDS);
+        }
+
+        let last_idx = self.len();
+        self.len += 1;
+        self.set(last_idx, element)
+    }
+}
+
+impl<T> Vector<T>
+where
+    T: BorshSerialize + BorshDeserialize,
+{
+    fn deserialize_element(raw_element: &[u8]) -> T {
+        T::try_from_slice(&raw_element).unwrap_or_else(|_| env::panic(ERR_ELEMENT_DESERIALIZATION))
+    }
+
+    /// Returns the element by index or `None` if it is not present.
+    pub fn get(&self, index: u32) -> Option<&T> {
+        let entry = self.cache.get(index).get_or_init(|| {
+            let storage_bytes = env::storage_read(&self.index_to_lookup_key(index));
+            let value = storage_bytes.as_deref().map(Self::deserialize_element);
+            CacheEntry::new_cached(value)
+        });
+        entry.value().as_ref()
+    }
+
+    /// Returns a mutable reference to the element at the `index` provided.
+    fn get_mut_inner(&mut self, index: u32) -> &mut CacheEntry<T> {
+        let index_to_lookup_key = self.index_to_lookup_key(index);
+        let entry = self.cache.get_mut(index);
+        entry.get_or_init(|| {
+            let storage_bytes = env::storage_read(&index_to_lookup_key);
+            let value = storage_bytes.as_deref().map(Self::deserialize_element);
+            CacheEntry::new_cached(value)
+        });
+        let entry = entry.get_mut().unwrap();
+        entry
```

Comment:
The entry is marked as modified when it's borrowed mutably. No need to do it twice and/or prematurely mark as modified in case this entry is never used

Reply:
I'd default to dirtying the entry as early as possible -- that way you can't have a bug. If the entry ends up not being used, I'd (as a rule of thumb) change the calling code to not do mutable lookup in the first place. I haven't look closely at all the calling contexts, so, if you feel the current approach is ok, it's fine to leave it as is.

Reply:
The value cannot be modified without marking it as dirty, so seems inconsistent to manually do it here and not do other places. 

In the current code, there is no way for this to be not marked as mutable, but I'd like to keep it as is if possible because the `replace` call can avoid marking as dirty if both values are `None`, so it's future-proofed and open to avoid unnecessary state writes.